// h5.modbus/index.js
exports.TcpConnection = require('./TcpConnection');
exports.TcpTransport = require('./TcpTransport');
exports.RtuTransport = require('./RtuTransport');
exports.AsciiTransport = require('./AsciiTransport');
exports.Transaction = require('./Transaction');
exports.Master = require('./Master');

var functions = require('./functions');

for (var fn in functions)
{
  exports[fn] = functions[fn];
}

exports.createMaster = function(options)
{
  var connection;
  var transport;
  
  switch (options.type)
  {
    case 'tcp':
      connection = new exports.TcpConnection(options);
      transport = new exports.TcpTransport(connection, options);
      break;

    case 'tcp-rtu':
      connection = new exports.TcpConnection(options);
      transport = new exports.RtuTransport(connection, options);
      break;

    case 'tcp-ascii':
      connection = new exports.TcpConnection(options);
      transport = new exports.AsciiTransport(connection, options);
      break;

    case 'serial-rtu':
      connection = new (require('./SerialConnection'))(options);
      transport = new exports.RtuTransport(connection, options);
      break;

    case 'serial-ascii':
      connection = new (require('./SerialConnection'))(options);
      transport = new exports.AsciiTransport(connection, options);
      break;

    default:
      throw new Error('Master of type [' + options.type + '] is not supported.');
  }
  
  return new exports.Master(connection, transport, options);
};
